server/tilesource/geospatial.py

```diff
@@ -0,0 +1,128 @@
+#############################################################################
+
+import gdal
+import osr
+import mapnik
+from PIL import Image
+import math
+
+from .base import FileTileSource
+
+try:
+    import girder
+    from girder import logger
+    from .base import GirderTileSource
+except ImportError:
+    girder = None
+    import logging as logger
+    logger.getLogger().setLevel(logger.INFO)
+
+
+class MapnikTileSource(FileTileSource):
+    name = 'mapnik'
+
+    def __init__(self, path, **kwargs):
+        super(MapnikTileSource, self).__init__(path, **kwargs)
+        self.dataset = gdal.Open(self._getLargeImagePath())
+
+        self.sizeX = self.dataset.RasterXSize
+        self.sizeY = self.dataset.RasterYSize
+        self.tileWidth = 256
+        self.tileHeight = 256
+        # # Fair assumption
+        self.levels = 20
+
+        # Earth circumference for web mercator
+        self.circumference = 20037508.342789244
+
+    def _getProj4String(self):
+        """Returns proj4 string for the given dataset"""
+        wkt = self.dataset.GetProjection()
+        proj = osr.SpatialReference()
+        proj.ImportFromWkt(wkt)
+
+        return proj.ExportToProj4()
+
+    def getPixelSize(self, zoom):
+        """Returns an approximate pixel size for a given zoom level"""
+
+        equatorLatitude = 0
+        # For simplification we assume everything is equator latitude
+        pixelSize = self.circumference * math.cos(equatorLatitude) / 2 ** (zoom + 7)
+
+        return pixelSize
+
+    def getTileCorners(self, z, x, y, tileSize=256):
+        """Returns bounds of a tile for a given x,y,z index"""
+
+        numberOfTiles = 2.0 ** (z * 2)
+        pixelSize = self.getPixelSize(z)
+        circumference = self.circumference
+        xmin = x / numberOfTiles * (circumference * 2 ** (z + 1)) - circumference
+        ymin = y / numberOfTiles * (circumference * 2 ** (z + 1)) - circumference
+        xmax = xmin + pixelSize * tileSize
+        ymax = ymin + pixelSize * tileSize
+
+        return xmin, ymin, xmax, ymax
+
+    def addStyle(self, m):
+        """Attaches raster style option to mapnik raster layer"""
+        style = mapnik.Style()
+        rule = mapnik.Rule()
+        sym = mapnik.RasterSymbolizer()
+        rule.symbols.append(sym)
+        style.rules.append(rule)
+        m.append_style('Raster Style', style)
+        lyr = mapnik.Layer('layer')
+        lyr.srs = self._getProj4String()
+        lyr.datasource = mapnik.Gdal(base='',
+                                     file=self._getLargeImagePath(),
+                                     band=-1)
+        lyr.styles.append('Raster Style')
+        m.layers.append(lyr)
+
+    @staticmethod
+    def getMercatorProjection():
+        """Returns proj4 string for mercator projection"""
+        srs = osr.SpatialReference()
+        srs.ImportFromEPSG(3857)
+
+        return srs.ExportToProj4()
+
+    def getTile(self, x, y, z, **kwargs):
```

Comment:
Wrap with `@methodcache` to allow tile caching.  By default, this uses all of `kwargs` as part of the cache hashing, so this will do the right thing if you process style information.